File: android/wear/src/main/java/com/kiters/wear/session/LiveSessionUploadState.kt
package com.kiters.wear.session

data class PendingLiveRecord(
    val jumpM: Double? = null,
    val airS: Double? = null,
    val speedKmh: Double? = null,
    val distKm: Double? = null,
) {
    val isEmpty: Boolean
        get() = jumpM == null && airS == null && speedKmh == null && distKm == null
}

data class StartUploadSnapshot(
    val isCurrent: Boolean,
    val sessId: Int?,
    val startInFlight: Boolean,
)

class LiveSessionUploadState {
    var activeSessionId: String? = null
        private set
    var serverSessId: Int? = null
        private set
    var isStartUploadInFlight: Boolean = false
        private set
    val trackPoints: MutableList<List<Int>> = mutableListOf()

    private var lastPingTimeMs: Long? = null
    private var lastTrackTimeMs: Long? = null
    private var lastStartAttemptMs: Long? = null
    private var sessionBestJumpM = 0.0
    private var sessionBestAirS = 0.0
    private var sessionBestSpeedKmh = 0.0
    private var sessionBestDistKm = 0.0
    private var pendingRecord = PendingLiveRecord()

    fun reset(sessionId: String?) {
        activeSessionId = sessionId
        serverSessId = null
        isStartUploadInFlight = false
        lastPingTimeMs = null
        lastTrackTimeMs = null
        lastStartAttemptMs = null
        trackPoints.clear()
        sessionBestJumpM = 0.0
        sessionBestAirS = 0.0
        sessionBestSpeedKmh = 0.0
        sessionBestDistKm = 0.0
        pendingRecord = PendingLiveRecord()
    }

    fun beginStartAttempt(sessionId: String, nowMs: Long, lat: Double, lng: Double): Boolean {
        if (activeSessionId != sessionId || serverSessId != null || isStartUploadInFlight) return false
        val lastStart = lastStartAttemptMs
        if (lastStart != null && nowMs - lastStart < 5_000L) return false

        lastStartAttemptMs = nowMs
        isStartUploadInFlight = true
        lastPingTimeMs = nowMs
        lastTrackTimeMs = nowMs
        appendTrackPoint(lat, lng)
        return true
    }

    fun acceptStart(sessionId: String, sessId: Int): Boolean {
        if (activeSessionId != sessionId) return false
        serverSessId = sessId
        isStartUploadInFlight = false
        return true
    }

    fun failStart(sessionId: String) {
        if (activeSessionId == sessionId) isStartUploadInFlight = false
    }

    fun appendTrackIfDue(nowMs: Long, lat: Double, lng: Double): Boolean {
        val lastTrack = lastTrackTimeMs ?: return false
        if (nowMs - lastTrack < 5_000L) return false
        appendTrackPoint(lat, lng)
        lastTrackTimeMs = nowMs
        return true
    }

    fun shouldPing(nowMs: Long): Boolean {
        val lastPing = lastPingTimeMs ?: return false
        if (nowMs - lastPing < 10_000L) return false
        lastPingTimeMs = nowMs
        return true
    }

    val pingBestJumpM: Double?
        get() = if (sessionBestJumpM > 0.0) sessionBestJumpM else null

    fun speedRecordIfImproved(kmh: Double): Double? {
        if (kmh <= sessionBestSpeedKmh + 1.0) return null
        sessionBestSpeedKmh = kmh
        return kmh
    }

    fun distanceRecordIfImproved(distKm: Double): Double? {
        if (distKm <= sessionBestDistKm + 0.1) return null
        sessionBestDistKm = distKm
        return distKm
    }

    fun jumpRecordIfImproved(heightM: Double, airS: Double): PendingLiveRecord? {
        val jumpBetter = heightM > sessionBestJumpM
        val airBetter = airS > sessionBestAirS
        if (!jumpBetter && !airBetter) return null

        if (jumpBetter) sessionBestJumpM = heightM
        if (airBetter) sessionBestAirS = airS
        return PendingLiveRecord(
            jumpM = if (jumpBetter) heightM else null,
            airS = if (airBetter) airS else null,
        )
    }

    fun enqueue(record: PendingLiveRecord) {
        pendingRecord = PendingLiveRecord(
            jumpM = maxOrCurrent(pendingRecord.jumpM, record.jumpM),
            airS = maxOrCurrent(pendingRecord.airS, record.airS),
            speedKmh = maxOrCurrent(pendingRecord.speedKmh, record.speedKmh),
            distKm = maxOrCurrent(pendingRecord.distKm, record.distKm),
        )
    }

    fun claimPendingRecord(sessionId: String): PendingLiveRecord? {
        if (activeSessionId != sessionId || pendingRecord.isEmpty) return null
        val record = pendingRecord
        pendingRecord = PendingLiveRecord()
        return record
    }

    fun startSnapshot(sessionId: String): StartUploadSnapshot =
        StartUploadSnapshot(
            isCurrent = activeSessionId == sessionId,
            sessId = serverSessId,
            startInFlight = isStartUploadInFlight,
        )

    private fun appendTrackPoint(lat: Double, lng: Double) {
        val compact = listOf((lat * 1e4).toInt(), (lng * 1e4).toInt())
        if (trackPoints.lastOrNull() != compact) trackPoints.add(compact)
    }

    private fun maxOrCurrent(current: Double?, next: Double?): Double? =
        when {
            next == null -> current
            current == null -> next
            else -> maxOf(current, next)
        }
}

File: Kiters/Kiters Watch App/Services/LiveSessionUploadState.swift
import Foundation

public struct PendingLiveRecord: Equatable {
    public var jumpM: Double?
    public var airS: Double?
    public var speedKmh: Double?
    public var distKm: Double?

    public init(jumpM: Double? = nil, airS: Double? = nil, speedKmh: Double? = nil, distKm: Double? = nil) {
        self.jumpM = jumpM
        self.airS = airS
        self.speedKmh = speedKmh
        self.distKm = distKm
    }

    public var isEmpty: Bool {
        jumpM == nil && airS == nil && speedKmh == nil && distKm == nil
    }
}

public struct StartUploadSnapshot: Equatable {
    public let isCurrent: Bool
    public let sessId: Int?
    public let startInFlight: Bool
}

public final class LiveSessionUploadState {
    public private(set) var activeSessionId: String?
    public private(set) var serverSessId: Int?
    public private(set) var isStartUploadInFlight = false
    public private(set) var trackPoints: [[Int]] = []

    private var lastPingTime: Date?
    private var lastTrackTime: Date?
    private var lastStartAttempt: Date?
    private var sessionBestJumpM: Double = 0
    private var sessionBestAirS: Double = 0
    private var sessionBestSpeedKmh: Double = 0
    private var sessionBestDistKm: Double = 0
    private var pendingRecord = PendingLiveRecord()

    public init() {}

    public func reset(sessionId: String?) {
        activeSessionId = sessionId
        serverSessId = nil
        isStartUploadInFlight = false
        lastPingTime = nil
        lastTrackTime = nil
        lastStartAttempt = nil
        trackPoints = []
        sessionBestJumpM = 0
        sessionBestAirS = 0
        sessionBestSpeedKmh = 0
        sessionBestDistKm = 0
        pendingRecord = PendingLiveRecord()
    }

    public func beginStartAttempt(sessionId: String, now: Date, lat: Double, lng: Double) -> Bool {
        guard activeSessionId == sessionId, serverSessId == nil, !isStartUploadInFlight else { return false }
        if let lastStartAttempt, now.timeIntervalSince(lastStartAttempt) < 5 { return false }

        lastStartAttempt = now
        isStartUploadInFlight = true
        lastPingTime = now
        lastTrackTime = now
        appendTrackPoint(lat: lat, lng: lng)
        return true
    }

    public func acceptStart(sessionId: String, sessId: Int) -> Bool {
        guard activeSessionId == sessionId else { return false }
        serverSessId = sessId
        isStartUploadInFlight = false
        return true
    }

    public func failStart(sessionId: String) {
        guard activeSessionId == sessionId else { return }
        isStartUploadInFlight = false
    }

    public func appendTrackIfDue(now: Date, lat: Double, lng: Double) -> Bool {
        guard let lastTrackTime, now.timeIntervalSince(lastTrackTime) >= 5 else { return false }
        appendTrackPoint(lat: lat, lng: lng)
        self.lastTrackTime = now
        return true
    }

    public func shouldPing(now: Date) -> Bool {
        guard let lastPingTime, now.timeIntervalSince(lastPingTime) >= 10 else { return false }
        self.lastPingTime = now
        return true
    }

    public var pingBestJumpM: Double? { sessionBestJumpM > 0 ? sessionBestJumpM : nil }
    public var currentServerSessId: Int? { serverSessId }

    public func speedRecordIfImproved(kmh: Double) -> Double? {
        guard kmh > sessionBestSpeedKmh + 1 else { return nil }
        sessionBestSpeedKmh = kmh
        return kmh
    }

    public func distanceRecordIfImproved(distKm: Double) -> Double? {
        guard distKm > sessionBestDistKm + 0.1 else { return nil }
        sessionBestDistKm = distKm
        return distKm
    }

    public func jumpRecordIfImproved(heightM: Double, airS: Double) -> PendingLiveRecord? {
        let jumpBetter = heightM > sessionBestJumpM
        let airBetter = airS > sessionBestAirS
        guard jumpBetter || airBetter else { return nil }

        if jumpBetter { sessionBestJumpM = heightM }
        if airBetter { sessionBestAirS = airS }
        return PendingLiveRecord(
            jumpM: jumpBetter ? heightM : nil,
            airS: airBetter ? airS : nil
        )
    }

    public func enqueue(_ record: PendingLiveRecord) {
        if let jumpM = record.jumpM {
            pendingRecord.jumpM = max(pendingRecord.jumpM ?? 0, jumpM)
        }
        if let airS = record.airS {
            pendingRecord.airS = max(pendingRecord.airS ?? 0, airS)
        }
        if let speedKmh = record.speedKmh {
            pendingRecord.speedKmh = max(pendingRecord.speedKmh ?? 0, speedKmh)
        }
        if let distKm = record.distKm {
            pendingRecord.distKm = max(pendingRecord.distKm ?? 0, distKm)
        }
    }

    public func claimPendingRecord(sessionId: String) -> PendingLiveRecord? {
        guard activeSessionId == sessionId, !pendingRecord.isEmpty else { return nil }
        let record = pendingRecord
        pendingRecord = PendingLiveRecord()
        return record
    }

    public func startSnapshot(sessionId: String) -> StartUploadSnapshot {
        StartUploadSnapshot(
            isCurrent: activeSessionId == sessionId,
            sessId: serverSessId,
            startInFlight: isStartUploadInFlight
        )
    }

    private func appendTrackPoint(lat: Double, lng: Double) {
        let compact = [Int(lat * 1e4), Int(lng * 1e4)]
        if trackPoints.last != compact {
            trackPoints.append(compact)
        }
    }
}
